[RFT] kirkwood: add support for kernel 5.10

mans0n opened the pull request:
kirkwood: rename files directory
kirkwood: copy files to kernel 5.10
kirkwood: refresh patches for kernel 5.10
kirkwood: update config for kernel 5.10
kirkwood: set testing kernel version to 5.10

Tested on Pogoplug V4.

mans0n committed:
kirkwood: rename files-5.4 to files

Move local DTS files from "files-5.4" to "files" directory so kernel 5.10
can use it.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
kirkwood: copy files to kernel 5.10

Copy config and patches to kernel 5.10.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
kirkwood: refresh patches for kernel 5.10

Refresh patches to make them apply to kernel 5.10.
The removed patch has been merged upstream.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
kirkwood: update config for kernel 5.10

Update config with make kernel_oldconfig.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n committed:
kirkwood: set testing kernel version to 5.10

Tested on Pogoplug V4.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

adschm commented on target/linux/kirkwood/config-5.10:
```diff
@@ -222,19 +230,21 @@ CONFIG_PINCTRL_SX150X=y
+# CONFIG_RTC_DRV_RV3032 is not set
 CONFIG_RTC_I2C_AND_SPI=y
 CONFIG_RTC_MC146818_LIB=y
-CONFIG_SATA_PMP=y
```
Wasn't there something special about this SATA_PMP=y? I only have a faded memory about this ...

CHKDSK88 replied:
It's required for some NAS devices.

mans0n replied:
I can manually add it back to the config file, but it disappears again when I run `kernel_oldconfig`.
My approach to fix it is removing the disabled SATA_PMP from the generic config. Do you think it will work?

adschm replied:
Well, I'd say either we add the missing symbol which is the reason for it being removed, or we add a package just enabling it to this target and then add that to DEFAULT_PACKAGES. Will that work or am I overlooking something (maybe I'll have some time to look up the old discussion, but probably not ...)

adschm replied:
Ah, sorry, didn't see your related new patch. Unfortunately, I cannot really judge whether this is proper.

Apart from that:

> This increases compressed kernel size by 4KB on kirkwood.

This is measured compared to the version that removed it from target config, I assume?
Because actually, I'd argue that it was present in 5.4, so the reference should be having it enabled in 5.10, too, so no size change at all?

mans0n replied:
Right, but in this case it also affects other targets (that use SATA but have not enabled SATA_PMP explicitly) so I wanted to make sure that it doesn't increase the kernel size that much. Maybe it would be better to test it on another target like oxnas?

CHKDSK88 replied:
This commit make problem with our config:
https://github.com/torvalds/linux/commit/bd322af15ce970737c9b71fe2f368fea63ae70b9

It causes, that `CONFIG_SATA_PMP` depends on `CONFIG_SATA_HOST`.

We can do a little hack: revert this commit for kirkwood target. But I'm not sure if that is best approach.

bobafetthotmail replied:
> > This increases compressed kernel size by 4KB on kirkwood.
> 
> This is measured compared to the version that removed it from target config, I assume?

it increases size for all other targets (apart from kirkwood and mvebu that have this enabled) already 

that's the reason you didn't merge the commit https://github.com/openwrt/openwrt/pull/2979/commits/6970cc4745924555be4da29f2e3bdc5962de32a4 that added support for that back in june 2020 https://github.com/openwrt/openwrt/pull/2979#issuecomment-643601777

Is there a way to make that  
KCONFIG:= \
	CONFIG_SATA_PMP=y \
	CONFIG_ATA

dependant on target somehow? like for example (won't work but explains what I mean)

IF TARGET_KIRKWOOD THEN 
KCONFIG:= \
	CONFIG_SATA_PMP=y \
	CONFIG_ATA
ELSE
KCONFIG:= \
	CONFIG_SATA_PMP=y \
	CONFIG_ATA

because that would solve the issue, and we could enable this only on the targets that actually can use it (kirkwood and mvebu)

adschm replied:
I see three options here:
1. do not disable in generic config (i.e. mans0n's commit): kernel enables it by default, so why not do it as well (and it's only 4 kiB)
2. do not set it in generic config, but disable it in the kconfig for all targets currently not using it (will that survive refresh?)
3. have a package for it again, and add that package to DEFAULT_PACKAGES for the relevant targets. (why did we remove that in the first place again?)

I personally would not make it even more complicated by adding target-dependent switches for the packages.

Anyway, I think we should separate this discussion. I'd consider to just add the symbol back to kirkwood kconfig manually to get the kernel bump done. With that solved, we could further discuss the PMP issue.

Since changes seem rather small on this target, I consider just merging it in the current state plus named adjustments?

bobafetthotmail replied:
It isn't a package. When I added it I just placed in kernel config and at the time the kernel config also had sata support included, so everything was good.

Then someone else had to drop stuff from kernel because kernel size limitations on the two kirkwood-based routers, and we left it in the kernel config bcause while this was a bit hacky it still worked, until that commit that clears the hacky options.

CONFIG_SATA_PMP Is a compile option for kernel libata, or **ata-core** kmod in OpenWrt. If it's enabled the **ata-core** will be built with port multiplier support.

That's why I asked if it was possible to add a target-specific KCONFIG to **ata-core**.

adschm replied:
> It isn't a package. When I added it I just placed in kernel config and at the time the kernel config also had sata support included, so everything was good.

But creating a meta-package that will just set it to =y will provide the same effect, and still be nicely configurable.

mans0n replied:
Sorry, I can't really judge what is the best option here (though I prefer option 1, but other tiny targets might think differently), and I don't have enough spare time to investigate this.

> Anyway, I think we should separate this discussion. I'd consider to just add the symbol back to kirkwood kconfig manually to get the kernel bump done. With that solved, we could further discuss the PMP issue.

I agree. That will make things more easier for now. I would like to open a separate PR for it myself, but as I said probably I wouldn't be able to handle that.

> 1. do not disable in generic config (i.e. mans0n's commit): kernel enables it by default, so why not do it as well (and it's only 4 kiB)

It sounds weird, but I also found out that the *uncompressed* kernel size is exactly the same whether the option is enabled or not.

mans0n replied:
Finally I got some time to contemplate this and I came up with another idea. #4292

mans0n committed:
kernel: enable SATA_PMP in kernel 5.10

SATA_PMP option is no longer exposed when no SATA host driver is enabled
since upstream linux commit bd322af15ce9 ("ata: make SATA_PMP option
selectable only if any SATA host driver is enabled").
So removing the disabled symbol from generic config will enable the feature
only for devices and targets that use SATA host drivers.

This increases compressed kernel size by 4KB on kirkwood.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Rebased due to 0621b23efb4f and added SATA_PMP patch.

mans0n commented:
Kernel with ALL_KMODS builds well, and I found no regression during run-test. I think it is ready for broader tests.

adschm commented:
I will pick the first patch in the meantime.

adschm commented:
I get prompts for:
Marvell 88E6xxx Ethernet switch fabric support (NET_DSA_MV88E6XXX) [Y/n/m/?] y
  Switch Global 2 Registers support (NET_DSA_MV88E6XXX_GLOBAL2) [Y/n/?] y
    PTP support for Marvell 88E6xxx (NET_DSA_MV88E6XXX_PTP) [N/y/?] (NEW)

Intel PXA2XX I2C Slave comms support (I2C_PXA_SLAVE) [N/y/?] (NEW)

NET_DSA_MV88E6XXX_PTP is removed in your refresh patch ...

adschm commented:
> I get prompts for:
> Marvell 88E6xxx Ethernet switch fabric support (NET_DSA_MV88E6XXX) [Y/n/m/?] y
> Switch Global 2 Registers support (NET_DSA_MV88E6XXX_GLOBAL2) [Y/n/?] y
> PTP support for Marvell 88E6xxx (NET_DSA_MV88E6XXX_PTP) [N/y/?] (NEW)
> 
> Intel PXA2XX I2C Slave comms support (I2C_PXA_SLAVE) [N/y/?] (NEW)
> 
> NET_DSA_MV88E6XXX_PTP is removed in your refresh patch ...

Maybe I got it now ...

adschm commented:
I've merged the kernel and left the PMP issue unresolved as discussed. Please submit the corresponding patch in a new PR so we can further discuss this issue.

mans0n commented:
Finally I got some time to contemplate this and I came up with another idea. #4292
